File: Java-Kotlin/src/main/java/jianzhi/jz16.java
package jianzhi;

import common.datastructure.ListNode;

public class jz16 {

    public ListNode Merge(ListNode list1, ListNode list2) {
        if (list1 == null && list2 == null) {
            return null;
        }

        if (list1 == null) {
            return list2;
        }

        if (list2 == null) {
            return list1;
        }

        ListNode dummy = new ListNode(0);
        ListNode runner = dummy;
        ListNode p1Runner = list1;
        ListNode p2Runner = list2;
        while (p1Runner != null && p2Runner != null) {
            if (p1Runner.val < p2Runner.val) {
                runner.next = p1Runner;
                p1Runner = p1Runner.next;
            } else {
                runner.next = p2Runner;
                p2Runner = p2Runner.next;
            }
            runner = runner.next;
        }

        if (p1Runner != null) {
            runner.next = p1Runner;
        }

        if (p2Runner != null) {
            runner.next = p2Runner;
        }

        return dummy.next;
    }
}

File: Go-Solutions/nowcoder/jz16.go
package nowcoder

func Mergejz16(pHead1 *ListNode, pHead2 *ListNode) *ListNode {
	if pHead1 == nil && pHead2 == nil {
		return nil
	}
	if pHead1 == nil {
		return pHead2
	}
	if pHead2 == nil {
		return pHead1
	}

	// write code here
	dummy := &ListNode{Val: 0, Next: nil}
	runner := dummy

	p1Runner, p2Runner := pHead1, pHead2
	for p1Runner != nil && p2Runner != nil {
		if p1Runner.Val < p2Runner.Val {
			runner.Next = p1Runner
			p1Runner = p1Runner.Next
		} else {
			runner.Next = p2Runner
			p2Runner = p2Runner.Next
		}
		runner = runner.Next
	}
	if p1Runner != nil {
		runner.Next = p1Runner
	}
	if p2Runner != nil {
		runner.Next = p2Runner
	}
	return dummy.Next
}
